Deprecate the 'vectorizeOnly' iterator

bradcray opened the pull request:
This PR deprecates the `vectorizeOnly` iterator since we prefer
`foreach` loops now.  Or, more accurately, it associates a
deprecation warning with the flag/pragma that we use to call
out the `vectorizeOnly()` iterator because using the normal
`deprecated` keyword resulted in triplicate deprecation warnings
for each loop for reasons I couldn't determine easily (best guess...
something about the way the compiler rewrote the code caused
it?)

Updated existing modules and tests to use `foreach` rather
than `vectorizeOnly` in the obvious ways.

Two pause points here were that `foreach` loops don't support
task intents yet, which means that a loop in LCALS that relied
on task intents couldn't be converted and now generates the
warning.  Elliot says this isn't a concern since we're not running
those variants and didn't see benefits from them anyway.
Similarly, I futurized one test that applied task intents to a
vectorizeOnly() loop, associating it with our open issue.

Where we typically deprecate aggressively after the warning has
been in for a release, in this case, we may want to wait until
`foreach` loops do support intents so that nothing is lost...?

Resolves #19238

bmcdonald3 reviewed (approved):
Looks good. 

I think the wording "in favor of using" is a little strange, and might instead go with something like "'vectorizeOnly()' is deprecated, consider using a 'foreach' loop instead" or something like that, but I won't put up a fight on the wording of it and think it is fine as is.

dlongnecke-cray commented on test/deprecated/vectorizeOnly.good:
```diff
@@ -0,0 +1,4 @@
+vectorizeOnly.chpl:1: warning: 'vectorizeOnly()' is deprecated in favor of using 'foreach' loops
```
It feels slightly confusing to me that 3 warnings are generated for one call, here. However I'm not so sure there's anything we can do about it.

bradcray replied:
Yeah, I just (re?) noticed that this afternoon as well, and was going to look into it a bit more.  I'm guessing it's something that the compiler is doing that may be hard to un-do, but I'll check.

bradcray replied:
I haven't been able to (easily) figure out what's going on.  Looking at the IR in the log files, it looks like there's only one call and that it's only getting invoked once.  Also, if I put a similar deprecation tag on a user iterator, even one with overloads, I get just a single message.  Odd.

All that said, the compiler has some special code to deal with vectorizeOnly() iterators, so (a) it could easily be that that's causing it in some way that I'm not picking up on, and (b) since @bmcdonald3 wanted me to update the message anyway... I could have the compiler generate the deprecation warning which only results in a single warning per loop rather than 3.

dlongnecke-cray commented on test/performance/vectorization/vectorizeOnly/vectorizeOnlyWithIntents.bad:
```diff
@@ -0,0 +1,6 @@
+vectorizeOnlyWithIntents.chpl:5: syntax error: near 'with'
```
I guess I'm surprised because I would have hoped that `foreach` would just copy the `forall` syntax productions. Maybe after concluding the "unify the intents" discussion we could also unify the grammar rules for most loop forms? Or maybe someone who touches the grammar frequently could unify them now and replace the errors with something else besides "syntax error"? (me)

bradcray replied:
I think the history here was that `for` loops haven't traditionally supported intents and `foreach` based its implementation more on `for` than `forall`?  But we are talking about adding intents for `for` loops, which I think(?) has generally been gaining favor (see https://github.com/chapel-lang/chapel/issues/17857).  And even if we didn't support them for `for` loops, I think it makes sense to for `foreach` given that they're parallel ( #18500 ).

If you were to put a nicer user error together, that would be nice.  That said, it's been in this state for awhile now and hasn't really bothered anybody that I'm ware of.

dlongnecke-cray reviewed (approved):
LGTM, had some questions for learning purposes.

bradcray commented:
@dlongnecke-cray / @bmcdonald3 : I think I've addressed both of your comments, although made some pretty significant changes in the code to do so.  Specifically, I:

* moved the deprecation warning from the module code's use of `deprecated` to a compiler warning to get rid of the triplicate error messages that I couldn't explain well (yet which were bugging me the more I thought about them)
* updated the wording while doing so
* changed the interface of the routine that prints the warning (the only place in the compiler that currently checks for the flag indicating it's a vectorizeOnly() iterator so that we'd get good source locations in the warnings
* updated the documentation to explicitly note that the iterator is deprecated
* updated the test to try  more loop forms
* updated the output of both the deprecation test and LCALS to indicate the new error message and its frequency

I'm running one more round of tests on this, but if that comes back clean, I'll merge based on the initial reviews and hope you can do a post-merge review tomorrow?

bradcray commented:
@ronawho: I ended up taking a different tack with this PR after puzzling over getting too many deprecation warnings in my first attempt and so would be curious to have you do a post-merge review on this since it involves tests that you'd added to test a pragma that propagated vectorization hints.  Specifically, I'll be curious whether the implication is that there's more code we could remove from the compiler than I was guessing when we remove vectorizeOnly() (I was imagining that most of it was reused by `foreach` but that could be a misassumption...  @e-kayrakli and @mppf may have thoughts on that as well).

ronawho commented:
I think this is fine.

> Specifically, I'll be curious whether the implication is that there's more code we could remove from the compiler than I was guessing when we remove vectorizeOnly

I'm not sure. I don't really remember much about this work offhand (somewhat because it was a while ago, but moreso because we didn't really see any benefit from it and I haven't worked on it since then.) I'd have to look at https://github.com/chapel-lang/chapel/pull/2201 and maybe parts of https://github.com/chapel-lang/chapel/pull/1490 to remember what all was added and then see how it morphed over time with `foreach` to guess what could be removed when `vectorizeOnly` is removed.
